### PR TITLE
feat: `value_unit_pair` module

### PR DESCRIPTION
**Pull request by aaraney**
Add `ValueUnitPair` and `ListUnitPair` classes. The former captures the concept of a value and a unit (i.e. `42[meters]`) while the latter is an extension of the concept but for a list of values.

Why do we need this? BMI modules: soil freeze thaw, soil moisture profile, lgar, and CFE all use this format in their configuration files. These classes provide a way for capturing the expressed concept as pydantic models. This will enable adding the aforementioned modules to `ngen.config`.

## Notes

- There are a few PRs that I am in the process of staging that will result in a `ngen.config` version bump, but this alone does not constitute a version bump.

### CODE REVIEW

File: python/ngen_conf/src/ngen/config/init_config/value_unit_pair.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+from typing import TYPE_CHECKING, Any, Generic, List, Optional, Type, TypeVar, Union
+
+from pydantic import validator
+from pydantic.generics import GenericModel
+from typing_extensions import Self, override
+
+if TYPE_CHECKING:
+    from pydantic.typing import AbstractSetIntStr, DictStrAny, MappingIntStrAny
+
+T = TypeVar("T")
```
Comment (hellkite500) [inline python/ngen_conf/src/ngen/config/init_config/value_unit_pair.py:11]: I, for one, would appreciate some context around these type vars -- what are they abstractly representing in their usage below?  A good comment string could be helpful to bring that context to the module.

File: python/ngen_conf/src/ngen/config/init_config/value_unit_pair.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+V = TypeVar("V")
+U = TypeVar("U")
+
+
+_VAL_UNIT_RE = re.compile(r"(.*)\[(.*)\]")
```
Comment (hellkite500) [inline python/ngen_conf/src/ngen/config/init_config/value_unit_pair.py:16]: Not that I think we should now, but may want to consider more restrictive regex here in the future...especially considering captures which have spaces, non-standard characters, ect.  This may be impossible to generalize and too far out of scope though -- just a passing thought.

File: python/ngen_conf/src/ngen/config/init_config/value_unit_pair.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    @override
+    @classmethod
+    def validate(cls: Type[Self], value: Any) -> Self:
+        if isinstance(value, ValueUnitPair):
+            return value
+
+        if isinstance(value, dict):
+            value = f"{value['value']}[{value['unit']}]"
+
+        if not isinstance(value, str):
```
Comment (hellkite500) [inline python/ngen_conf/src/ngen/config/init_config/value_unit_pair.py:32]: this extra type check here seems irrelevant and kinda confusing...should the previous `if` not simply return value after creating the format string, eliminating this check?

**Review by hellkite500 — CHANGES_REQUESTED** [review]

File: python/ngen_conf/src/ngen/config/init_config/value_unit_pair.py (hunk 3 above)
Comment (aaraney) [reply python/ngen_conf/src/ngen/config/init_config/value_unit_pair.py:32]: Yeah thanks for pointing this out. I will add a comment here. There is a lot of pydantic magic happening in this method that if you don't know _how_ pydantic works under the hook then you would be pretty understandably lost...

`validate` is special to pydantic. Its what actually validates and coerces some value into a type _or partially_ does that.  pydantic will iterate over the functions its gets from that `__get_validators__` dundermethod, passing the nth result to the nth+1 validator and so on. These are functions are shaped like the `validate` class method above. So, in this case, if at this point we don't have a string that we can do something with, we just return it and pydantic will pass it to the next validator method in the chain.

File: python/ngen_conf/src/ngen/config/init_config/value_unit_pair.py (hunk 2 above)
Comment (aaraney) [reply python/ngen_conf/src/ngen/config/init_config/value_unit_pair.py:16]: Yeah I agree. I didn't and don't know at this point in time what appropriate constraints to enforce here, so I left it a little open. My thinking was we could always refine this if needed.

File: python/ngen_conf/src/ngen/config/init_config/value_unit_pair.py (hunk 1 above)
Comment (aaraney) [reply python/ngen_conf/src/ngen/config/init_config/value_unit_pair.py:11]: Yeah that is fair. Ill add a docstring explaining their purpose a bit.

File: python/ngen_conf/src/ngen/config/init_config/value_unit_pair.py (hunk 3 above)
Comment (hellkite500) [reply python/ngen_conf/src/ngen/config/init_config/value_unit_pair.py:32]: But under what circumstance can we get to this point without having a string we can do something with?

Comment (aaraney) [reply python/ngen_conf/src/ngen/config/init_config/value_unit_pair.py:32]: Yeah now that I look at it closer and poke around with it more, I think you are right. Great catch. There are several other edge cases that ive glossed over. I am addressing those and will request a re-review.

**Review by hellkite500 — APPROVED** [review]

**Comment by aaraney** [comment]
Thanks, @hellkite500!
